src/main/scala/util/Repeater.scala

```diff
@@ -2,38 +2,41 @@
 
 package freechips.rocketchip.util
 
-import Chisel._
+import chisel3._
+import chisel3.util.{Decoupled, DecoupledIO}
 
-// A Repeater passes it's input to it's output, unless repeat is asserted.
+// A Repeater passes its input to its output, unless repeat is asserted.
 // When repeat is asserted, the Repeater copies the input and repeats it next cycle.
 class Repeater[T <: Data](gen: T) extends Module
 {
-  val io = new Bundle {
-    val repeat = Bool(INPUT)
-    val full = Bool(OUTPUT)
-    val enq = Decoupled(gen).flip
-    val deq = Decoupled(gen)
-  }
+  val typ = chiselTypeOf(gen)
```

Comment:
I think you should have just let it using 'gen'. That will make it the caller's job to chiselTypeOf() on the constructor argument, which is probably a better API.

Reply:
While I agree your thinking is more in line with the chisel3 philosophy, I thought it was more practical to maintain API compatibility, rather than having to change all the call sites (some of which have nothing to do with this PR).